util/dex/dex.go

```diff
@@ -20,9 +21,10 @@ var messageRe = regexp.MustCompile(`<p>(.*)([\s\S]*?)<\/p>`)
 // with the external issuer URL muxed to the same path configured in server.go. In other words, if
 // Argo CD API server wants to proxy requests at /api/dex, then the dex config yaml issuer URL should
 // also be /api/dex (e.g. issuer: https://argocd.example.com/api/dex)
-func NewDexHTTPReverseProxy(serverAddr string) func(writer http.ResponseWriter, request *http.Request) {
+func NewDexHTTPReverseProxy(serverAddr string, baseHRef string) func(writer http.ResponseWriter, request *http.Request) {
 	target, err := url.Parse(serverAddr)
 	errors.CheckError(err)
+	target.Path = baseHRef
```

Comment:
Hm, you sure that this is correct? Shouldn't it append the Dex context as well?

Reply:
I've tested  it using Dex + Github Oauth app.

Just in case here is a list of steps:

1. Created Github oauth app that uses `http://localhost:9000/argocd/api/dex/callback` as a callback url
2. Configured dex sso integration:
```yaml
apiVersion: v1
data:
  dex.config: |
    connectors:
      - type: github
        id: github
        name: GitHub
        config:
          clientID: xxxx
          clientSecret: xxxxx
          orgs:
          - name: argoproj
  url: http://localhost:9000/argocd
kind: ConfigMap
metadata:
  labels:
    app.kubernetes.io/part-of: argocd
  name: argocd-cm
```

3. Run dex, argocd-server, and redis with the config below locally:
```
ARGOCD_E2E_DISABLE_AUTH=false goreman run api-server redis dex
nginx -g 'daemon off;' -c nginx-reverse-proxy.conf                                                                                                                                                                             
```
nginx-reverse-proxy.conf                                                                                                                                                                             
```
events {
  worker_connections  1024;
}

http {
    server {
        listen       9000;
        location /argocd/ {
            rewrite /argocd/(.*) /$1  break;
            proxy_pass http://localhost:8080;
        }

        # buffering should be disabled for api/v1/stream/applications to support chunked response
        proxy_buffering off;
    }
}
```

Access http://localhost:9000/argocd

Without the changes SSO the Argo CD sends the request to http://localhost:5556/api/dex/auth and SSO  flow fails with the 404 error.  With the changes, the request is sent to http://localhost:5556/argocd/api/dex/auth and successfully completes.

It would be nice to have e2e test for it, but I would rather test it manually now, create ticket and implement test after 1.5 release

Reply:
What I ment was, given `serverAddr` is `http://example.com/api/dex` and `baseHRef` is `/argocd` - before this change, `target.Path` was `/api/dex`, with this patch it will be `/argocd` - and I _think_ it should be `/argocd/api/dex` instead. Or is my assumption inherintly wrong here? While it will work with `/argocd` as path, I would feel it not being 100% correct, because we only want Dex requests to be routed through the reverse proxy.

Reply:
The `serverAddr` does not have a path. It is usually `http://argocd-dex-server:5556`. I think this is causing confusion. 

The `/api/dex` is in the proxied request. After adding `/argocd` to the target URL the reverse proxy sends `/api/dex` to `/argocd/api/dex`